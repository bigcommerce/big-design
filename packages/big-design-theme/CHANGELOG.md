# Change Log

## 1.3.0

### Minor Changes

- 042d470: Change the documentation and code to use Source Sans 3 instead of Source Sans Pro. Source Sans Pro is licensed by Adobe now which means we need to the utilize the open license with Source Sans 3.

## 1.2.0

### Minor Changes

- e323932: added negative values to spacing definitions

## 1.1.0

### Minor Changes

- f8665f3: Added wide breakpoint to theme definition to best adapt responsive features to wide screens

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.2](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@1.0.1...@bigcommerce/big-design-theme@1.0.2) (2024-08-06)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [1.0.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@1.0.0...@bigcommerce/big-design-theme@1.0.1) (2024-06-11)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [1.0.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.21.1...@bigcommerce/big-design-theme@1.0.0) (2024-04-17)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.21.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.21.0...@bigcommerce/big-design-theme@0.21.1) (2024-04-09)

### Bug Fixes

- typescript build output with wrong imports ([a812497](https://github.com/bigcommerce/big-design/commit/a81249798e20c10754c938b2b3d3dcc95d81aa0c))

# [0.21.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.20.1...@bigcommerce/big-design-theme@0.21.0) (2024-04-02)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.20.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.20.0...@bigcommerce/big-design-theme@0.20.1) (2024-01-23)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.20.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.19.2...@bigcommerce/big-design-theme@0.20.0) (2023-10-09)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.19.2](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.19.1...@bigcommerce/big-design-theme@0.19.2) (2023-06-14)

### Bug Fixes

- declaration tsconfig for theme publishing ([4789b8b](https://github.com/bigcommerce/big-design/commit/4789b8b35d3fbfbd42d0312de66cd1b48bef0f39))

## [0.19.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.19.0...@bigcommerce/big-design-theme@0.19.1) (2023-06-09)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.19.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.18.0...@bigcommerce/big-design-theme@0.19.0) (2023-04-05)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.18.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.17.0...@bigcommerce/big-design-theme@0.18.0) (2023-01-09)

### Features

- convert to react@18 only ([#1044](https://github.com/bigcommerce/big-design/issues/1044)) ([06c4697](https://github.com/bigcommerce/big-design/commit/06c469721bc06d1f872bed6bf5a46e6b568644e5))

### BREAKING CHANGES

- Requires `react@18` or higher.

# [0.17.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.16.1...@bigcommerce/big-design-theme@0.17.0) (2022-09-27)

### Features

- **deps:** react@18 support ([51c688b](https://github.com/bigcommerce/big-design/commit/51c688b282277964eb01c1ed67c5c27cc0bc4c3e))

## [0.16.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.16.0...@bigcommerce/big-design-theme@0.16.1) (2022-09-27)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.16.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.15.1-alpha.0...@bigcommerce/big-design-theme@0.16.0) (2022-08-17)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.15.1-alpha.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.15.0...@bigcommerce/big-design-theme@0.15.1-alpha.0) (2022-08-10)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.15.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.14.1...@bigcommerce/big-design-theme@0.15.0) (2022-07-06)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.14.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.14.0...@bigcommerce/big-design-theme@0.14.1) (2022-06-10)

### Features

- **theme:** re-export hideVisually from polished ([e102e6d](https://github.com/bigcommerce/big-design/commit/e102e6dfc6d28b311556946d3e90d2323885713f))

# [0.14.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.13.2-alpha.1...@bigcommerce/big-design-theme@0.14.0) (2022-05-11)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.13.2-alpha.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.13.2-alpha.0...@bigcommerce/big-design-theme@0.13.2-alpha.1) (2022-03-22)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.13.2-alpha.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.13.1...@bigcommerce/big-design-theme@0.13.2-alpha.0) (2021-10-28)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.13.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.13.0...@bigcommerce/big-design-theme@0.13.1) (2021-09-03)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.13.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.13.0-alpha.2...@bigcommerce/big-design-theme@0.13.0) (2021-08-04)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.13.0-alpha.2](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.13.0-alpha.1...@bigcommerce/big-design-theme@0.13.0-alpha.2) (2021-08-02)

### chore

- **deps:** upgrade to styled-components v5 ([#566](https://github.com/bigcommerce/big-design/issues/566)) ([dd83711](https://github.com/bigcommerce/big-design/commit/dd83711797eb8aaa7a8406eebf1383116bff2420))

### Features

- **theme:** add listReset helper ([#570](https://github.com/bigcommerce/big-design/issues/570)) ([385b142](https://github.com/bigcommerce/big-design/commit/385b1426ee39171343070f9b31ad4e31d834b272))

### BREAKING CHANGES

- **deps:** You will need to update to styled-components v5. In addition,
  you will need to import the base fonts in your <head> element.
  See the "Getting Started" page or README.md for an example.

# [0.13.0-alpha.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.13.0-alpha.0...@bigcommerce/big-design-theme@0.13.0-alpha.1) (2021-07-06)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.13.0-alpha.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.12.1...@bigcommerce/big-design-theme@0.13.0-alpha.0) (2021-06-10)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.12.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.12.0...@bigcommerce/big-design-theme@0.12.1) (2021-03-29)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.12.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.11.0...@bigcommerce/big-design-theme@0.12.0) (2021-03-08)

### Bug Fixes

- **docs:** Specify styled-components version in the README ([#490](https://github.com/bigcommerce/big-design/issues/490)) ([c92403e](https://github.com/bigcommerce/big-design/commit/c92403edf8e29bf1afd3c9c163c37be6ffe17a6c))

### Features

- **theme:** add currentColor, inherit, and transparent props to colors ([#504](https://github.com/bigcommerce/big-design/issues/504)) ([3ef9b18](https://github.com/bigcommerce/big-design/commit/3ef9b1899ccef4c282f2f293e8fdd34d204e0420))

# [0.11.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.10.1...@bigcommerce/big-design-theme@0.11.0) (2020-11-12)

### Bug Fixes

- webpack 5 compatibility ([#466](https://github.com/bigcommerce/big-design/issues/466)) ([7e66238](https://github.com/bigcommerce/big-design/commit/7e66238ca42cb27d91b1a80cc9e1c8014808e27b))
- **component:** updates progress circle to use em instead of rem due to error in Safari 14 ([#472](https://github.com/bigcommerce/big-design/issues/472)) ([3bedcc1](https://github.com/bigcommerce/big-design/commit/3bedcc1be8e5f5a8c9c63d634a1c567dd3016002))

## [0.10.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.10.0...@bigcommerce/big-design-theme@0.10.1) (2020-08-31)

### Features

- **theme:** add DefaultTheme extension types ([cdfeb49](https://github.com/bigcommerce/big-design/commit/cdfeb493973cce48ed92e30951be2b8bc87ed194))

# [0.10.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.9.0...@bigcommerce/big-design-theme@0.10.0) (2020-06-16)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.9.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.8.0...@bigcommerce/big-design-theme@0.9.0) (2020-04-29)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.8.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.7.0...@bigcommerce/big-design-theme@0.8.0) (2020-04-20)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.7.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.6.0...@bigcommerce/big-design-theme@0.7.0) (2020-03-25)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# [0.6.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.5.0...@bigcommerce/big-design-theme@0.6.0) (2020-02-21)

### Features

- **component:** add Alerts, InlineAlerts, and Message components ([#340](https://github.com/bigcommerce/big-design/issues/340)) ([d541276](https://github.com/bigcommerce/big-design/commit/d54127603fba47b46cb35c3db4caf53ab24bafc3))

# [0.5.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.4.0...@bigcommerce/big-design-theme@0.5.0) (2020-02-06)

### Bug Fixes

- add react-hooks eslint and fix missing deps/issues ([#330](https://github.com/bigcommerce/big-design/issues/330)) ([da3fbd6](https://github.com/bigcommerce/big-design/commit/da3fbd68181e98e43a95de7fce9956be91afc9b8))

# [0.4.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.3.1...@bigcommerce/big-design-theme@0.4.0) (2019-12-09)

### Features

- **component:** add tab trap for modals and dialogs ([dcacbf9](https://github.com/bigcommerce/big-design/commit/dcacbf96a38bef1134e2a8dcbd986f6362e0e2b7))

## [0.3.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.3.0...@bigcommerce/big-design-theme@0.3.1) (2019-11-12)

### Features

- **all:** better tree shaking ([c0998a7](https://github.com/bigcommerce/big-design/commit/c0998a7))

# [0.3.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.2.0...@bigcommerce/big-design-theme@0.3.0) (2019-11-11)

### Features

- **all:** bump styled-components peerDependency version ([fd89fa3](https://github.com/bigcommerce/big-design/commit/fd89fa3))
- **component:** tooltip for dropdown item ([#228](https://github.com/bigcommerce/big-design/issues/228)) ([4e5fc50](https://github.com/bigcommerce/big-design/commit/4e5fc50))

### BREAKING CHANGES

- **all:** bumped peer dependency of styled-components to `^4.3.0`

# [0.2.0](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.1.2...@bigcommerce/big-design-theme@0.2.0) (2019-10-29)

**Note:** Version bump only for package @bigcommerce/big-design-theme

## [0.1.2](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.1.1...@bigcommerce/big-design-theme@0.1.2) (2019-08-19)

### Bug Fixes

- **component:** add missing z-index to components ([#176](https://github.com/bigcommerce/big-design/issues/176)) ([800d2fc](https://github.com/bigcommerce/big-design/commit/800d2fc))

### BREAKING CHANGES

- **component:** `theme.zIndex` no longer has `theme.zIndex.dropdown`

## [0.1.1](https://github.com/bigcommerce/big-design/compare/@bigcommerce/big-design-theme@0.1.0...@bigcommerce/big-design-theme@0.1.1) (2019-08-15)

**Note:** Version bump only for package @bigcommerce/big-design-theme

# 0.1.0 (2019-08-13)

### Bug Fixes

- **theme:** export missing properties ([f439df0](https://github.com/bigcommerce/big-design/commit/f439df0))
- **theme:** issue with font sizes in createTheme ([#135](https://github.com/bigcommerce/big-design/issues/135)) ([313ff46](https://github.com/bigcommerce/big-design/commit/313ff46))

### Features

- **theme:** initial package setup ([f734dfb](https://github.com/bigcommerce/big-design/commit/f734dfb))
- **theme:** rename elevation to shadow ([659c6ad](https://github.com/bigcommerce/big-design/commit/659c6ad))

### BREAKING CHANGES

- **theme:** `elevation` is now called `shadow`
