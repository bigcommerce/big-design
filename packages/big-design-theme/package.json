{
  "name": "@bigcommerce/big-design-theme",
  "version": "1.1.0",
  "sideEffects": false,
  "main": "src/index.ts",
  "license": "(MIT AND CC-BY-3.0)",
  "author": "BigCommerce Inc",
  "repository": {
    "type": "git",
    "url": "https://github.com/bigcommerce/big-design.git",
    "directory": "packages/big-design-theme"
  },
  "scripts": {
    "build": "pnpm run build:cjs && pnpm run build:es && pnpm run build:dt",
    "build:cjs": "NODE_ENV=production BABEL_ENV=cjs babel --extensions \".ts,.tsx\" ./src --out-dir ./dist/cjs",
    "build:es": "NODE_ENV=production BABEL_ENV=es babel --extensions \".ts,.tsx\" ./src --out-dir ./dist/es",
    "build:dt": "tsc -p tsconfig.declarations.json --emitDeclarationOnly",
    "lint": "eslint . --max-warnings 0",
    "prepack": "big-design-prepack",
    "postpack": "big-design-postpack",
    "test": "jest",
    "test:watch": "pnpm run test --watch",
    "typecheck": "tsc --noEmit"
  },
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public",
    "main": "dist/cjs/index.js",
    "module": "dist/es/index.js",
    "typings": "dist/index.d.ts"
  },
  "prettier": "@bigcommerce/eslint-config/prettier",
  "dependencies": {
    "@babel/runtime": "^7.26.0",
    "polished": "^4.3.1"
  },
  "peerDependencies": {
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "styled-components": "^5.3.5"
  },
  "devDependencies": {
    "@babel/cli": "^7.25.7",
    "@babel/core": "^7.26.0",
    "@babel/plugin-transform-runtime": "^7.25.9",
    "@babel/preset-env": "^7.26.0",
    "@babel/preset-react": "^7.26.3",
    "@babel/preset-typescript": "^7.26.0",
    "@bigcommerce/configs": "workspace:^",
    "@bigcommerce/pack": "workspace:^",
    "@styled/typescript-styled-plugin": "^1.0.1",
    "@swc/core": "^1.10.4",
    "@swc/jest": "^0.2.37",
    "@swc/plugin-styled-components": "^6.0.2",
    "@types/jest": "^29.5.14",
    "@types/styled-components": "^5.1.34",
    "babel-jest": "^29.0.2",
    "babel-plugin-styled-components": "^2.0.7",
    "jest": "^29.7.0",
    "jest-environment-jsdom": "^29.4.1",
    "jest-fail-on-console": "^3.3.1",
    "jest-styled-components": "^7.1.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "styled-components": "^5.3.11",
    "typescript": "^5.6.3"
  }
}
