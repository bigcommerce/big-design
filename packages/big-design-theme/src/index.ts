import { createHelpers, Helpers } from './helpers';
import { ThemeOptions, themeOptions } from './options';
import { Border, BorderRadius, createBorder, createBorderRadius } from './system/border';
import { Breakpoints, breakpoints, BreakpointValues, breakpointValues } from './system/breakpoints';
import { Colors, colors } from './system/colors';
import * as keyframes from './system/keyframes';
import { createLineHeight, LineHeight } from './system/line-height';
import { Shadow, shadow } from './system/shadow';
import { createSpacing, Spacing } from './system/spacing';
import { createTypography, Typography } from './system/typography';
import { ZIndex, zIndex } from './system/z-index';

export * from './helpers';
export * from './system';

export interface ThemeInterface {
  border: Border;
  borderRadius: BorderRadius;
  breakpointValues: BreakpointValues;
  breakpoints: Breakpoints;
  colors: Colors;
  helpers: Helpers;
  keyframes: typeof keyframes;
  lineHeight: LineHeight;
  shadow: Shadow;
  spacing: Spacing;
  typography: Typography;
  zIndex: ZIndex;
}

export const createTheme = (customOptions: Partial<ThemeOptions> = {}): ThemeInterface => {
  themeOptions.setOptions(customOptions);

  return {
    border: createBorder(),
    borderRadius: createBorderRadius(),
    breakpointValues,
    breakpoints,
    colors,
    helpers: createHelpers(),
    keyframes,
    lineHeight: createLineHeight(),
    shadow,
    spacing: createSpacing(),
    typography: createTypography(),
    zIndex,
  };
};

export const theme: ThemeInterface = createTheme();
