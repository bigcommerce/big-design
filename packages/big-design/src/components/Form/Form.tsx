import React, { ComponentPropsWithoutRef, forwardRef, Ref } from 'react';

import { MarginProps } from '../../helpers';
import { typedMemo } from '../../utils';

import { StyledForm } from './styled';
import { FormContext } from './useFormContext';

interface PrivateProps {
  readonly forwardedRef: Ref<HTMLFormElement>;
}

export type FormProps = ComponentPropsWithoutRef<'form'> &
  MarginProps & {
    readonly fullWidth?: boolean;
  };

const StyleableForm: React.FC<PrivateProps & FormProps> = ({ forwardedRef, ...props }) => {
  return (
    <FormContext.Provider value={{ fullWidth: props.fullWidth }}>
      <StyledForm {...props} ref={forwardedRef} />
    </FormContext.Provider>
  );
};

export const Form = typedMemo(
  forwardRef<HTMLFormElement, FormProps>(({ className, style, ...props }, ref) => (
    <StyleableForm {...props} forwardedRef={ref} />
  )),
);
