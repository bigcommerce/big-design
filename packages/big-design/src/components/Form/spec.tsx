import { fireEvent, render, screen } from '@testing-library/react';
import React, { createRef } from 'react';
import 'jest-styled-components';

import { Fieldset } from '../Fieldset';
import { Input } from '../Input';

import { Form, FormGroup } from './index';

test('forwards ref', () => {
  const ref = createRef<HTMLFormElement>();

  render(<Form aria-label="Test form" ref={ref} />);

  const form = screen.getByRole('form');

  expect(form).toBe(ref.current);
});

test('calls onSubmit', async () => {
  const onSubmit = jest.fn();

  render(<Form aria-label="Test form" name="testForm" onSubmit={onSubmit} />);

  const form = await screen.findByRole('form');

  fireEvent.submit(form);

  expect(onSubmit).toHaveBeenCalled();
});

test('simple form render', () => {
  render(
    <Form aria-label="Test form">
      <Fieldset
        description="Minim velit quis aute adipisicing adipisicing do do exercitation cupidatat enim ex voluptate consequat labore."
        legend="Primary contact"
      >
        <FormGroup>
          <Input
            description="This is an example description for First Name"
            label="First Name"
            placeholder="Placeholder text"
          />
        </FormGroup>

        <FormGroup>
          <Input
            description="This is an example description for Last Name. Featuring a Left Icon."
            label="Middle Name"
            placeholder="Placeholder text"
          />
        </FormGroup>
      </Fieldset>
    </Form>,
  );

  const form = screen.getByRole('form');

  expect(form).toMatchSnapshot();
});

test('has margin props', () => {
  const { rerender } = render(<Form aria-label="Test form" />);

  const form = screen.getByRole('form');

  expect(form).not.toHaveStyle('margin: 1rem');

  rerender(<Form aria-label="Test form" margin="medium" />);

  expect(form).toHaveStyle('margin: 1rem');
});
