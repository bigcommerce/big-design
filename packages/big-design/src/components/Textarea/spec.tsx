import 'jest-styled-components';
import { theme } from '@bigcommerce/big-design-theme';
import { render, screen } from '@testing-library/react';
import React, { createRef } from 'react';

import { warning } from '../../utils';
import { FormControlDescription, FormControlError, FormControlLabel, FormGroup } from '../Form';

import { Textarea } from './index';

const basicBorderStyle = `1px solid ${theme.colors.secondary30}`;
const errorBorderStyle = `1px solid ${theme.colors.danger40}`;

test('forwards ref', () => {
  const ref = createRef<HTMLTextAreaElement>();

  render(<Textarea ref={ref} />);

  const textarea = screen.getByRole('textbox');

  expect(ref.current).toBe(textarea);
});

test('renders an texarea tag', () => {
  render(<Textarea />);

  expect(screen.getByRole('textbox')).toBeInTheDocument();
});

test('renders an textarea with matched label', () => {
  render(<Textarea label="Test Label" />);

  // This one checks for matching id and htmlFor
  expect(screen.getByLabelText('Test Label')).toBeInTheDocument();
});

test('create unique ids if not provided', async () => {
  render(
    <>
      <FormGroup>
        <Textarea data-testid="item1" label="Test Label" />
      </FormGroup>
      <FormGroup>
        <Textarea data-testid="item2" label="Test Label" />
      </FormGroup>
    </>,
  );

  const item1 = await screen.findByTestId('item1');
  const item2 = await screen.findByTestId('item2');

  expect(item1).toBeInTheDocument();
  expect(item2).toBeInTheDocument();
  expect(item1.id).not.toBe(item2.id);
});

test('respects provided id', async () => {
  render(<Textarea id="test" label="Test Label" />);

  const textarea = await screen.findByLabelText<HTMLTextAreaElement>('Test Label');

  expect(textarea.id).toBe('test');
});

test('matches label htmlFor with id provided', async () => {
  render(<Textarea id="test" label="Test Label" />);

  const label = await screen.findByText<HTMLLabelElement>('Test Label');

  expect(label.htmlFor).toBe('test');
});

test('respects provided labelId', async () => {
  render(<Textarea label="Test Label" labelId="test" />);

  const label = await screen.findByText<HTMLLabelElement>('Test Label');

  expect(label.id).toBe('test');
});

test('renders a description', () => {
  const descriptionText = 'This is a description';

  render(<Textarea description={descriptionText} />);

  expect(screen.getByText(descriptionText)).toBeInTheDocument();
});

test('renders an error', () => {
  const errorText = 'This is an error';

  render(
    <FormGroup>
      <Textarea data-testid="test-textarea" error={errorText} />
    </FormGroup>,
  );

  const textarea = screen.getByTestId('test-textarea');

  expect(screen.getByText(errorText)).toBeInTheDocument();
  expect(textarea).toHaveStyleRule('border', errorBorderStyle);
});

test('accepts a Label Component', () => {
  const CustomLabel = (
    <FormControlLabel>
      This is a custom Label
      <a data-testid="test" href="#">
        has a url
      </a>
    </FormControlLabel>
  );

  render(<Textarea label={CustomLabel} />);

  expect(screen.getByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Label Components', () => {
  const NotALabel = (
    <div>
      This is a not custom Label Component
      <a data-testid="test" href="#">
        has a url
      </a>
    </div>
  );

  render(<Textarea label={NotALabel} />);

  expect(screen.queryByTestId('test')).not.toBeInTheDocument();
});

test('accepts a Description Component', () => {
  const CustomDescription = (
    <FormControlDescription>
      This is a custom Description
      <a data-testid="test" href="#">
        has a url
      </a>
    </FormControlDescription>
  );

  render(<Textarea description={CustomDescription} />);

  expect(screen.getByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Description Components', () => {
  const NotADescription = (
    <div>
      This is a not custom description
      <a data-testid="test" href="#">
        has a url
      </a>
    </div>
  );

  render(<Textarea description={NotADescription} />);

  expect(screen.queryByTestId('test')).not.toBeInTheDocument();
});

test('accepts an Error Component', () => {
  const CustomError = (
    <FormControlError>
      This is a custom Error Component
      <a data-testid="test" href="#">
        has a url
      </a>
    </FormControlError>
  );

  render(
    <FormGroup>
      <Textarea data-testid="test-textarea" error={CustomError} />
    </FormGroup>,
  );

  const textarea = screen.getByTestId('test-textarea');

  expect(screen.getByTestId('test')).toBeInTheDocument();
  expect(textarea).toHaveStyleRule('border', errorBorderStyle);
});

test('does not accept non-Error Components', () => {
  const NotAnError = (
    <div>
      This is a not a custom error component
      <a data-testid="test" href="#">
        has a url
      </a>
    </div>
  );

  render(
    <FormGroup>
      <Textarea data-testid="test-textarea" error={NotAnError} />
    </FormGroup>,
  );

  const textarea = screen.getByTestId('test-textarea');

  expect(screen.queryByTestId('test')).not.toBeInTheDocument();
  expect(textarea).toHaveStyleRule('border', basicBorderStyle);
});

describe('error does not show when invalid type', () => {
  test('single element', () => {
    const error = <div data-testid="err">Error</div>;

    render(
      <FormGroup>
        <Textarea data-testid="test-textarea" error={error} />
      </FormGroup>,
    );

    const textarea = screen.getByTestId('test-textarea');

    expect(warning).toHaveBeenCalledTimes(1);
    expect(screen.queryByTestId('err')).not.toBeInTheDocument();
    expect(textarea).toHaveStyleRule('border', basicBorderStyle);
  });

  test('array of elements', () => {
    const errors = [
      'Error',
      <FormControlError key="1">Error</FormControlError>,
      <div data-testid="err" key="2">
        Error
      </div>,
    ];

    render(
      <FormGroup>
        <Textarea data-testid="test-textarea" error={errors} />
      </FormGroup>,
    );

    const textarea = screen.getByTestId('test-textarea');

    expect(warning).toHaveBeenCalledTimes(1);
    expect(screen.queryByTestId('err')).not.toBeInTheDocument();
    expect(textarea).toHaveStyleRule('border', errorBorderStyle);
  });
});

test('accepts valid row property', async () => {
  const rows = 3;

  render(<Textarea data-testid="test-rows" rows={rows} />);

  const textarea = await screen.findByTestId('test-rows');

  expect(textarea).toHaveAttribute('rows', `${rows}`);
});

test('does not accept invalid row property', async () => {
  const rows = 9;

  // @ts-expect-error negative test
  render(<Textarea data-testid="test-rows" rows={rows} />);

  const textarea = await screen.findByTestId('test-rows');

  expect(textarea.getAttribute('rows')).not.toBe(`${rows}`);
});

test('renders and accepts resize property', () => {
  const { rerender } = render(<Textarea data-testid="test-textarea" resize={true} />);
  const textarea = screen.getByTestId('test-textarea');

  expect(textarea).toHaveStyle('resize: vertical');

  rerender(<Textarea resize={false} />);

  expect(textarea).toHaveStyle('resize: none');
});

test('renders all together', () => {
  render(
    <Textarea description="This is a description" label="This is a label" resize={true} rows={3} />,
  );

  expect(screen.getByRole('textbox')).toMatchSnapshot();
});

describe('error shows when an array of strings', () => {
  test('empty array', () => {
    render(
      <FormGroup>
        <Textarea data-testid="test-textarea" error={[]} />
      </FormGroup>,
    );

    const textarea = screen.getByTestId('test-textarea');

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
    expect(textarea).toHaveStyleRule('border', basicBorderStyle);
  });

  test('array with empty strings', () => {
    render(
      <FormGroup>
        <Textarea data-testid="test-textarea" error={['', '']} />
      </FormGroup>,
    );

    const textarea = screen.getByTestId('test-textarea');

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
    expect(textarea).toHaveStyleRule('border', basicBorderStyle);
  });

  test('array with valid strings', () => {
    const errors = ['Error 0', 'Error 1'];

    render(
      <FormGroup>
        <Textarea data-testid="test-textarea" error={errors} />
      </FormGroup>,
    );

    const textarea = screen.getByTestId('test-textarea');

    expect(textarea).toHaveStyleRule('border', errorBorderStyle);

    errors.forEach((error) => expect(screen.getByText(error)).toBeInTheDocument());
  });
});
