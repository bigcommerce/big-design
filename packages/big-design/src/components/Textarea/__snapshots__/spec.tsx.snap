// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders all together 1`] = `
.c0 {
  -webkit-transition: all 150ms ease-out;
  transition: all 150ms ease-out;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  background-color: #FFFFFF;
  border-radius: 0.25rem;
  box-sizing: border-box;
  color: #313440;
  line-height: 1.5rem;
  max-height: 14rem;
  padding: 0.25rem 0.75rem;
  width: 100%;
  resize: vertical;
  border: 1px solid #D9DCE9;
}

.c0:hover:not([disabled]) {
  border: 1px solid #B4BAD1;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0 4px #DBE3FE;
}

.c0[disabled] {
  background-color: #ECEEF5;
}

.c0::-webkit-input-placeholder {
  color: #B4BAD1;
  line-height: 1.5rem;
  font-size: 1rem;
}

.c0::-moz-placeholder {
  color: #B4BAD1;
  line-height: 1.5rem;
  font-size: 1rem;
}

.c0:-ms-input-placeholder {
  color: #B4BAD1;
  line-height: 1.5rem;
  font-size: 1rem;
}

.c0::placeholder {
  color: #B4BAD1;
  line-height: 1.5rem;
  font-size: 1rem;
}

<textarea
  class="c0"
  id=":rp:"
  rows="3"
/>
`;
